yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr

```diff
@@ -0,0 +1,419 @@
+            AccountContract::at(from.address).is_valid(Context::private(&mut context), message_field);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let from_balance = storage.balances.at(from.address);
+
+        decrement(from_balance, amount, from.address);
+
+        let selector = compute_selector("_reduce_total_supply(Field)");
+        let _void = context.call_public_function(context.this_address(), selector, [amount]);
+
+        1
+    }
+
+    /// SHOULD BE Internal ///
+
+    // We cannot do this from the constructor currently 
+    // Since this should be internal, for now, we ignore the safety checks of it, as they are 
+    // enforced by it being internal and only called from the constructor.
+    #[aztec(public)]
+    fn _initialize(
+        new_admin: AztecAddress,
+    ) {
+        let storage = Storage::init(Context::public(&mut context));
+        storage.admin.write(new_admin.address);
+        storage.minters.at(new_admin.address).write(1);
+    }
+
+    /// Internal ///
+
+    #[aztec(public)]
+    internal fn _increase_public_balance(
+        to: AztecAddress,
+        amount: Field,
+    ) {
+        let storage = Storage::init(Context::public(&mut context));
+        let new_balance = SafeU120::new(storage.public_balances.at(to.address).read()).add(SafeU120::new(amount));
+        storage.public_balances.at(to.address).write(new_balance.value as Field);
+    }
+
+    #[aztec(public)]
+    internal fn _reduce_total_supply(
+        amount: Field,
+    ) {
+        // Only to be called from burn.
+        let storage = Storage::init(Context::public(&mut context));
+        let new_supply = SafeU120::new(storage.total_supply.read()).sub(SafeU120::new(amount));
+        storage.total_supply.write(new_supply.value as Field);
+    }
+
+    /// Unconstrained /// 
+
+    unconstrained fn admin() -> Field {
+        let storage = Storage::init(Context::none());
+        storage.admin.read()
+    }
+
+    unconstrained fn is_minter(
+        minter: AztecAddress,
+    ) -> bool {
+        let storage = Storage::init(Context::none());
+        storage.minters.at(minter.address).read() as bool
+    }
+
+    unconstrained fn total_supply() -> Field {
+        let storage = Storage::init(Context::none());
+        storage.total_supply.read()
+    }
+
+    unconstrained fn balance_of_private(
+        owner: AztecAddress,
+    ) -> Field {
+        let storage = Storage::init(Context::none());
+        let owner_balance = storage.balances.at(owner.address);
+
+        balance_utils::get_balance(owner_balance)
+    }
+
+    unconstrained fn balance_of_public(
+        owner: AztecAddress,
+    ) -> Field {
+        let storage = Storage::init(Context::none());
+        storage.public_balances.at(owner.address).read()
+    }
+
+    // Below this point is the stuff of nightmares.
+    // This should ideally not be required. What do we do if vastly different types of preimages?
```

Comment:
Since we'll likely need to define many 'sizes' of `compute_note_hash_and_nullifier`, there should be a way for the note processor to distinguish between different note types. The ciphertexts probably need to contain some identifier for the note type.

> This should ideally not be required

I'm not sure I agree with this. Having the Noir Contract express how the private state database should be populated (as is done here) gives lots of control to developers. It took a lot of discussions to get to this pattern :')